nucypher/blockchain/eth/sol/source/contracts/PREApplication.sol

```diff
@@ -0,0 +1,652 @@
+
+    /**
+    * @notice Constructor sets address of token contract and parameters for staking
+    * @param _token T token contract
+    * @param _tStaking T token staking contract
+    * @param _hashAlgorithm Hashing algorithm
+    * @param _basePenalty Base for the penalty calculation
+    * @param _penaltyHistoryCoefficient Coefficient for calculating the penalty depending on the history
+    * @param _percentagePenaltyCoefficient Coefficient for calculating the percentage penalty
+    * @param _minAuthorization Amount of minimum allowable authorization
+    * @param _minOperatorSeconds Min amount of seconds while an operator can't be changed
+    * @param _rewardDuration Duration of one reward cycle
+    * @param _deauthorizationDuration Duration of decreasing authorization
+    */
+    constructor(
+        IERC20 _token,
+        IStaking _tStaking,
+        SignatureVerifier.HashAlgorithm _hashAlgorithm,
+        uint256 _basePenalty,
+        uint256 _penaltyHistoryCoefficient,
+        uint256 _percentagePenaltyCoefficient,
+        uint256 _minAuthorization,
+        uint256 _minOperatorSeconds,
+        uint256 _rewardDuration,
+        uint256 _deauthorizationDuration
+    )
+        Adjudicator(
+            _hashAlgorithm,
+            _basePenalty,
+            _penaltyHistoryCoefficient,
+            _percentagePenaltyCoefficient
+        )
+    {
+        require(
+            _rewardDuration != 0 &&
+            _tStaking.authorizedStake(address(this), address(this)) == 0 &&
+            _token.totalSupply() > 0,
+            "Wrong input parameters"
+        );
+        rewardDuration = _rewardDuration;
+        deauthorizationDuration = _deauthorizationDuration;
+        minAuthorization = _minAuthorization;
+        token = _token;
+        tStaking = _tStaking;
+        minOperatorSeconds = _minOperatorSeconds;
+    }
+
+    /**
+    * @dev Update reward for the specified staking provider
+    */
+    modifier updateReward(address _stakingProvider) {
+        updateRewardInternal(_stakingProvider);
+        _;
+    }
+
+    /**
+    * @dev Checks caller is T staking contract
+    */
+    modifier onlyStakingContract()
+    {
+        require(msg.sender == address(tStaking), "Caller must be the T staking contract");
+        _;
+    }
+
+    /**
+    * @dev Checks caller is a staking provider or stake owner
+    */
+    modifier onlyOwnerOrStakingProvider(address _stakingProvider)
+    {
+        require(isAuthorized(_stakingProvider), "Not owner or provider");
+        if (_stakingProvider != msg.sender) {
+            (address owner,,) = tStaking.rolesOf(_stakingProvider);
+            require(owner == msg.sender, "Not owner or provider");
+        }
+        _;
+    }
+
+    /**
+     * @notice Initialize function for using with OpenZeppelin proxy
+     */
+    function initialize() external initializer {
+        __Ownable_init();
+    }
+
+    //------------------------Reward------------------------------
```

Comment:
In line with my comment above, I think we should explore what can we abstract regarding rewards management to a common interface for Threshold apps.

Reply:
threshold-network/solidity-contracts#89